content/blog/2017/05/2017-05-12-kubernetes-journey-on-azure.adoc

```diff
@@ -0,0 +1,397 @@
+tools (`make`, `bats`, `jq` `gpg`, etc).  The more tools we used, the more
+issues appeared because of the different versions used.
+
+Another challenge that emerged when dealing with different environments was:
+how to manage environment's specific configurations (dev, prod, etc).
+
+Is it better to define different configuration files per environment? Perhaps,
+but this means code duplications, or using file templates that require more
+tools (`sed`, `jinja2`, `erb`). And more work, but less code to write and
+maintain.
+
+There isn't a golden rule we discovered, and the answer is probably in between.
+
+In any case, the good news is that `Jenkinsfile` provides an easy way to
+execute tasks from a docker image and an image can contain all the necessary
+tools in our environment, we can even use different docker images for each
+stage along the way.
+
+In the following example, I use the `my_env` Docker image. It contains all the
+tools needed to test, validate, and deploy changes.
+
+.Jenkinsfile
+[source,groovy]
+----
+pipeline{
+  agent {
+    docker{
+      image 'my_env:1.0'
+    }
+  }
+  options{
+    buildDiscarder(logRotator(numToKeepStr: '10'))
+    disableConcurrentBuilds()
+    timeout(time: 1, unit: 'HOURS')
+  }
+  triggers{
+    pollSCM('* * * * *')
+  }
+  stages{
+    stage('Init'){
+      steps{
+        // Init everything required to deploy our infra
+        sh 'make init'
+      }
+    }
+    stage('Test'){
+      steps{
+       // Run tests to validate changes
+       sh 'make test'
+      }
+    }
+    stage('Deploy'){
+      steps{
+       // Deploy changes in production
+       sh 'make deploy'
+      }
+    }
+  }
+  post{
+    always {
+      sh 'make notify'
+    }
+  }
+}
+----
+
+=== Secret credentials
+
+Managing secrets is a big subject and brings with it many different
+requirements which are very hard to fulfill.  For obvious reasons, we couldn't
+publish the credentials used within the infra project.  On the other side, we
+needed to keep track and share them. Particularly for the Jenkins node that
+deploys our cluster.  This means that we needed a way to encrypt or decrypt
+those credentials depending on permissions, environments, etc.  We analyzed two
+different approaches to handle this:
+
+  1. Storing secrets in a key management tool like https://azure.microsoft.com/en-us/services/key-vault/[Key Vault] or https://www.vaultproject.io/[Vault] and use them like a kubernetes "secret" type of resource. +
+    -> Unfortunately, these tools are not yet integrated in Kubernetes. But we may come back to this option later.
+    https://github.com/kubernetes/kubernetes/issues/10439[Kubernetes issue: 10439]
+
+  2. Publishing and encrypting using a public gpg key. +
+     This means that everybody can encrypt credentials for the infrastructure project but only the owner of the private key can decrypt credentials. +
+     This solution implies
+      * Scripting, as secrets need to be decrypted at deployment time.
```

Comment:
Weird standalone terms, and sentences (there's a period) without a verb. Either do full sentences, or something line

This solution implies two things:
* Scripting, as secrets need to be …, and
* Templates, as secrets